src/Swarm/Language/Typecheck.hs

```diff
@@ -452,6 +452,9 @@ infer s@(Syntax l t) = (`catchError` addLocToTypeErr s) $ case t of
     c2' <- maybe id ((`withBinding` Forall [] a) . lvVar) mx $ infer c2
     _ <- decomposeCmdTy (c2' ^. sType)
     return $ Syntax' l (SBind mx c1' c2') (c2' ^. sType)
+  SAnnotate c pty -> do
+    uty <- skolemize $ toU pty
+    check c uty
```

Comment:
This is the right idea but I think it has a couple issues:
1. The general idea is that the typechecker annotates the AST with types, but otherwise leaves the AST intact.  I think even in the annotation case we should leave the annotation there instead of stripping it off (for example, because we might later want to pretty-print the term in an error message and make sure we preserve the way the user wrote it; or even more relevant, for #7 we might want to later pretty-print the term to save it in a file, in which case it might not even typecheck any more if we got rid of the annotation).  So instead of just `check c uty` this should say something like `c' <- check c uty; return $ Syntax' l (SAnnotate c' pty) ...`
2. The second issue is that after skolemizing the type in order to check it, we need to re-generalize so that we don't end up later trying to match skolem variables against other types when the annotated thing is used.  In other words, the skolemized type should only be used internally to *check* that the annotated thing is valid with the given type annotation; it should not be used *externally* as the inferred type of the thing.  Actually it may be as simple as using `toU pty` as the external type.  So something like `return $ Syntax' l (SAnnotate c' pty) (toU pty)`.  I haven't tested this, however.

Reply:
Thank you! Again I immediately agree with your point 1 that it is not a good idea to erase type signatures at this point. I should have paid closer attention to the `SLam` and `SLet` cases to realize this. 

Regarding your point 2 I had to think and experiment a bit to get to a workable solution. Your proposal doesn't quite work, since `toU pty` is of type `UPolytype` where `UType` is needed. I think the actual problem is that I invoke `skolemize` to replace universally quantified type variables with instances that cannot unify with anything other than themselves; surely we have a problem if they escape to the external world. In think we should rather `instantiate` them with unification variables. 

However, if I implement it this way, I get strange errors about a Skolem variable escaping its scope under certain circumstances.  An example for this is the following expression:
```
((\x.x):(a->a)->(a->a))(\x:a->a.x)
```
I can evaluate it once and get the type `(a->a)->a->a`, while a second evaluation of this expression raises the error `Skolem variable a would escape its scope`. It seems to me that the reason for this problem lies not in the new type ascription but rather in how type annotations _inside_ a lambda expression are handled. As the corresponding constructor in the AST is `SLam LocVar (Maybe Type) (Syntax' ty)` and not something like `SLam LocVar (Maybe Polytype) (Syntax' ty)` a type annotation in this context is not polymorphic and free type variables seem to be interpreted as a kind of type constant. 

I would like to look into this problem a little longer to evaluate whether it makes sense to solve it in this merge request or enter a separate issue for it.